jmclient/jmclient/configure.py

```diff
@@ -730,6 +775,119 @@ def load_program_config(config_path="", bs=None, plugin_services=[]):
+            # our LightningRpc object:
+            global_singleton.config.set(lnchanconfig["section-name"],
+                    "lightning-rpc", lnchanconfig["clightning-location"])
+
+def start_ln(chaninfo, jm_ln_dir, brpc_net, last_digit=-1):
+    """ If a LN message channel is configured with the
+    # "bundled" setting, we start the packaged/customised
+    # lightning daemon with the required ports
+    # in the background at the startup of each script.
+    """
+    global_singleton.config.set(chaninfo["section-name"], "lightning-rpc",
+                    os.path.join(jm_ln_dir, brpc_net, "lightning-rpc"))
+
+    passthrough_port = str(chaninfo["passthrough-port"])
+    # find where our custom lightningd is:
+    lightningd_loc = os.path.join(sys.prefix, "bin")
+    if not os.path.exists(os.path.join(lightningd_loc, "lightningd")):
+        raise Exception("Can't find our custom lightningd")
+    # To get the location of the jmcl plugin, we need the location of this python file:
+    floc = os.path.dirname(os.path.abspath(__file__))
+    jmcl_loc = os.path.join(floc, "..", "..", "jmdaemon", "jmdaemon", "jmcl.py")
+    command = [os.path.join(lightningd_loc, "lightningd"), "--plugin="+jmcl_loc,
+               "--jmport="+passthrough_port, "--lightning-dir=" + jm_ln_dir]
+    # testing needs static values:
+    if brpc_net == "regtest":
+        fixed_key_str = "121212121212121212121212121212121212121212121212121212121212121" + str(last_digit)
+        command.append("--dev-force-privkey="+fixed_key_str)
+    # we need to create c-lightning's own config file, in lightningdir/config.
+    # This requires the bitcoin rpc config also:
+    brpc_host = global_singleton.config.get("BLOCKCHAIN", "rpc_host")
+    brpc_port = global_singleton.config.get("BLOCKCHAIN", "rpc_port")
+    brpc_user = global_singleton.config.get("BLOCKCHAIN", "rpc_user")
+    brpc_password = global_singleton.config.get("BLOCKCHAIN", "rpc_password")
+    ln_serving_port = chaninfo["lightning-port"]
+    lnconfiglines = ["bitcoin-rpcconnect=" + brpc_host,
+                     "bitcoin-rpcport=" + brpc_port,
+                     "bitcoin-rpcuser=" + brpc_user,
+                     "bitcoin-rpcpassword=" + brpc_password,
+                     "network=" + brpc_net]
+    if brpc_net == "regtest":
+        lnconfiglines += ["addr=127.0.0.1:" + str(ln_serving_port),
+                          "log-level=debug",
+                          "log-file=" + jm_ln_dir + "/log",
+                          "dev-fast-gossip",
+                          "dev-bitcoind-poll=2"]
+    else:
+        lnconfiglines += ["proxy=127.0.0.1:9050",
+                     "bind-addr=127.0.0.1:" + str(ln_serving_port),
+                     "addr=autotor:127.0.0.1:9051/torport=" + str(ln_serving_port),
+                     "always-use-proxy=true"]
+    with open(os.path.join(jm_ln_dir, "config"), "w") as f:
+        f.write("\n".join(lnconfiglines))
```

Comment:
If `start_ln()` is called at the startup of each script, can we avoid recreating the config file each time? This way a user could also modify his lightning config file without it being overwritten.

Reply:
Interesting thought. But I would tend to say no: first, a concrete reason: this `config` file for c-lightning is being fed configuration values from the user's `joinmarket.cfg` (Core RPC values for example), so we wouldn't want those to fail to carry over. But more fundamentally, if a user is letting Joinmarket control the installation and bundling it for them, they shouldn't need to change anything. If we want to the user to be able to configure anything else, we could add it to the lightning messaging section of joinmarket.cfg and pass it through as we do now for the RPC settings.

Reply:
I see, yeah adding them to `joinmarket.cfg` makes sense too, I think at the very least we should add a configurable Tor proxy value, such a small configuration change should not require a custom clightning installation.
Another example is the `announce-addr=myaddress.onion` thing I mentioned yesterday (I used it for the above coinjoin, so it seems to work nicely), it's kinda a big waste having to give up entirely the convenience of the bundled version just for little things like those.

Reply:
I see, I guessed you meant that. Could you just write exactly what fields need to be passed through from `joinmarket.cfg` to `config`, and what their defaults should be? Then I'll add them. Also a comment to put in about their purpose.

Reply:
Ok, I'll think about that.
It'd be nice if others, hopefully more well versed with clightning config, could chime in too.

Reply:
I mean, we can just add whatever you see a need for now, and then add others in future if people express the need for those.

Reply:
The quotes are from `man lightningd-config`:

#### proxy

> Set a socks proxy to use to connect to Tor nodes (or for all connections if always-use-proxy is set).  The port defaults to 9050 if not specified.

This would be the equivalent of the `socks5_host` and `socks5_port` that we have for IRC, or `onion_socks5_host` and `onion_socks5_port` that we have for Payjoin.

#### announce-addr

> Set an IP (v4 or v6) address or Tor address to announce; a Tor address is distinguished by ending in .onion. PORT defaults to 9735.

This is one of three options mentioned in the clightning [docs](https://lightning.readthedocs.io/TOR.html#tor-created-onion-address). This Tor stuff is very similar to Payjoin, we could reuse a lot here, to give a sketch:

```
# for c-lightning onion service creation.
# there is usually no need to change these settings
# the tor control configuration:
tor_control_host = 127.0.0.1
# or, to use a UNIX socket
# control_host = unix:/var/run/tor/control
tor_control_port = 9051
# the host serving the lightning onion service
lightning-host = 127.0.0.1
# Use a static onion address generated by Tor instead.
# This overrides the configuration above.
# Example: announce-addr=myaddress.onion
#announce-addr =
# this is the port serving lightning on your onion service, bundled or otherwise:
lightning-port = 9735
```

`lightning-port` can be reused for both `autotor` (default config) and `announce-addr`.
Example clightning config:

```
proxy={socks_host}:{socks_port}
announce-addr={myaddress.onion}:{lightning-port}
always-use-proxy=true
```

#### bitcoin-dir/bitcoin-cli

>  The name of bitcoin-cli executable to run.

> -datadir argument to supply to bitcoin-cli.

Not sure how relevant this are to our usage of clightning, having a non-default Bitcoin datadir/executable is pretty common, would it cause trouble?

#### tor-service-password

> Set a Tor control password, which may be needed for autotor: to authenticate to the Tor control port.

This seems a nice to have.

---

Just brainstorming the non-reachable taker mode:

We could add a flag like:

```
# to only enable outbound connections, no onion service created
#non-reachable = true
```

And this would generate a config with only:

```
proxy={socks_host}:{socks_port}
always-use-proxy=true
```

Reply:
>  bitcoin-dir/bitcoin-cli
> Not sure how relevant this are to our usage of clightning, having a non-default Bitcoin datadir/executable is pretty common, would it cause trouble?

Good point, I didn't think about that; we should definitely add this part.

> tor-service-password

Yes agreed.

> Just brainstorming the non-reachable taker mode:

I think I agree with all your comments here but clearly it's quite a bit of extra work to test it all.